deep_quoridor/src/agents/simple.py

```diff
@@ -1,61 +1,175 @@
+    # likely to place walls near themselves or their opponent, rather than somewhere else on the board.
+    wall_sigma: float = 0.5
+
+    # Future rewards are worth less. This makes the agent try to win quickly. Without this,
+    # once the agent knows it can win, it may oscillate between two moves instead of just winning.
+    discount_factor: float = 0.99
+
+
+def compute_wall_weight(wall: WallAction, position_1: np.ndarray, position_2: np.ndarray, sigma) -> float:
+    wall_position = np.array(wall.position)
+    d1 = np.linalg.norm(position_1 - wall_position)
+    d2 = np.linalg.norm(position_2 - wall_position)
+    min_distance = min(d1, d2)
+    weight = np.exp(-0.5 * (min_distance / sigma) ** 2)
+    return weight
+
+
+def sample_actions(game: Quoridor, n: int, wall_sigma: float | None = None) -> list[Action]:
+    """
+    Choose a sample of valid actions for the current player.
+
+    Starts by including all valid move actions, then adds a random sample of valid wall actions. May return less than
+    n actions if there are not enough valid actions.
     """
-    Sample a random sequence of actions for a given game. Stops early if the game terminates."""
-    action_sequence = []
-    total_reward = 0.0
-    reward_calculator = StepRewardCalculator(game, player, opponent)
-    while len(action_sequence) < max_path_length:
-        # For now, assume the other agent takes random actions.
-        valid_actions = game.get_valid_actions()
-        assert len(valid_actions) > 0, "No valid actions available... this shouldn't be possible."
+    actions = game.get_valid_move_actions()
+    if len(actions) > n:
+        actions = random.sample(actions, n)
+
+    if len(actions) < n:
+        num_wall_actions = n - len(actions)
+        wall_actions = game.get_valid_wall_actions()
+        if len(wall_actions) > num_wall_actions:
+            if wall_sigma is None:
+                wall_actions = random.sample(wall_actions, num_wall_actions)
+            else:
+                position_1 = np.array(game.board.get_player_position(Player.ONE))
+                position_2 = np.array(game.board.get_player_position(Player.TWO))
+                wall_weights = [compute_wall_weight(wall, position_1, position_2, wall_sigma) for wall in wall_actions]
+                wall_actions = np.random.choice(
+                    wall_actions,
+                    size=num_wall_actions,
+                    replace=False,
+                    p=wall_weights / np.sum(wall_weights),
+                )
+
+        actions.extend(wall_actions)
 
-        action = random.choice(valid_actions)
+    return actions
 
-        if game.get_current_player() == player:
-            action_sequence.append(action)
-            reward_calculator.before_step()
 
+def choose_action(
+    game: Quoridor,
+    player: Player,
+    opponent: Player,
+    max_depth: int,
+    branching_factor: int,
+    wall_sigma: float | None,
+    discount_factor: float,
+) -> float:
```

Comment:
I think this should be:

```suggestion
) -> tuple[Optional[Action], float]:
```